feat: add `toReversed` method to `array/fixed-endian-factory`

jangnathan opened the pull request:
Resolves #3159 

## Description

> What is the purpose of this pull request?

This pull request:
- Adds toReversed method to main.js
- Adds tests for toReversed method in new file test.to_reversed.js
- Adds two benchmark files for toReversed method, one measuring baseline performance and the other measuring performance as array length increases
- Added documentation and updated README.md file for toReversed method

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves #3159 

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

jangnathan committed:
test: add tests for toReverse method in array/fixed-endian-factory

jangnathan committed:
bench: add benchmarks for toReverse method in array/fixed-endian-factory

jangnathan committed:
docs: add documentation for toReverse method in array/fixed-endian-factory

ahmad-kashkoush commented on lib/node_modules/@stdlib/array/fixed-endian-factory/test/test.to_reversed.js:
```diff
@@ -0,0 +1,120 @@
+
+	t.strictEqual( instanceOf( out, ctor ), true, 'returns expected value' );
+	t.notEqual( out, arr, 'returns a new instance' );
+	t.strictEqual( out.length, expected.length, 'returns expected value' );
```
this test doesn't compare values of `expected` against values of `out`, it just compares the length.

jangnathan replied:
made changes

jangnathan committed:
fix: add missing test to toReverse method in array/fixed-endian-factory

jangnathan committed:
fix: resolve merge conflicts

kgryte commented on lib/node_modules/@stdlib/array/fixed-endian-factory/benchmark/benchmark.to_reversed.js:
```diff
@@ -0,0 +1,55 @@
+		}
+	}
+	b.toc();
+	if ( !( out instanceof Float64ArrayFE) ) {
```
```suggestion
	if ( !( out instanceof Float64ArrayFE ) ) {
```

kgryte commented on lib/node_modules/@stdlib/array/fixed-endian-factory/benchmark/benchmark.to_reversed.js:
```diff
@@ -0,0 +1,55 @@
+	}
+	b.toc();
+	if ( !( out instanceof Float64ArrayFE) ) {
+		b.fail( 'should return a Typed Array' );
```
```suggestion
		b.fail( 'should return a typed array' );
```

kgryte commented on lib/node_modules/@stdlib/array/fixed-endian-factory/benchmark/benchmark.to_reversed.length.js:
```diff
@@ -0,0 +1,101 @@
+* @returns {Function} benchmark function
+*/
+function createBenchmark( len ) {
+	var arr;
+
+	arr = new Float64ArrayFE( 'little-endian', zeroTo( len ) );
```
```suggestion
	var arr = new Float64ArrayFE( 'little-endian', zeroTo( len ) );
```

kgryte commented on lib/node_modules/@stdlib/array/fixed-endian-factory/benchmark/benchmark.to_reversed.length.js:
```diff
@@ -0,0 +1,101 @@
+		}
+		b.toc();
+		if ( !( out instanceof Float64ArrayFE ) ) {
+			b.fail( 'should return a Complex64Array' );
```
```suggestion
			b.fail( 'should return a typed array' );
```

kgryte commented on lib/node_modules/@stdlib/array/fixed-endian-factory/lib/main.js:
```diff
@@ -888,6 +888,56 @@ function factory( dtype ) { // eslint-disable-line max-lines-per-function, stdli
+	/**
+	* Returns a new typed array containing the elements in reversed order.
+	*
+	* @name toReversed
```
```suggestion
	* @private
	* @name toReversed
```

kgryte commented on lib/node_modules/@stdlib/array/fixed-endian-factory/lib/main.js:
```diff
@@ -888,6 +888,56 @@ function factory( dtype ) { // eslint-disable-line max-lines-per-function, stdli
 		return acc;
 	});
 
+	/**
```
This should be inserted such that methods are in alphabetical order.

kgryte commented on lib/node_modules/@stdlib/array/fixed-endian-factory/lib/main.js:
```diff
@@ -888,6 +888,56 @@ function factory( dtype ) { // eslint-disable-line max-lines-per-function, stdli
+	* @type {Function}
+	* @throws {TypeError} `this` is not a typed array
+	* @returns {TypedArray} reversed array
+	*
+	* @example
+	* var Float64ArrayFE = factory( 'float64' );
+	*
+	* var arr = new Float64ArrayFE( 'little-endian', [1.0, 2.0, 3.0] );
+	*
+	* var out = arr.toReversed();
+	* // returns <TypedArray>
+	*
+	* var v = out.get( 0 );
+	* // returns 1.0
+	*
+	* v = out.get( 1 );
+	* // returns 2.0
+	*
+	* v = out.get( 2 );
+	* // returns 3.0
```
```suggestion
```

We haven't been including examples in nested functions.

kgryte commented on lib/node_modules/@stdlib/array/fixed-endian-factory/lib/main.js:
```diff
@@ -888,6 +888,56 @@ function factory( dtype ) { // eslint-disable-line max-lines-per-function, stdli
+		buf = this._buffer;
+		obuf = out._buffer; // eslint-disable-line no-underscore-dangle
+		for ( i = 0; i < len; i++ ) {
+			v = buf[ GETTER ]( ( ( len - 1 ) * BYTES_PER_ELEMENT ) - ( i * BYTES_PER_ELEMENT ), this._isLE );
```
You're doing more arithmetic than is necessary.

```suggestion
			v = buf[ GETTER ]( ( len - i - 1 ) * BYTES_PER_ELEMENT, this._isLE );
```

kgryte commented on lib/node_modules/@stdlib/array/fixed-endian-factory/lib/main.js:
```diff
@@ -888,6 +888,56 @@ function factory( dtype ) { // eslint-disable-line max-lines-per-function, stdli
+		if ( !isTypedArray( this ) ) {
+			throw new TypeError( 'invalid invocation. `this` is not a typed array.' );
+		}
+
```
```suggestion
```

kgryte commented on lib/node_modules/@stdlib/array/fixed-endian-factory/lib/main.js:
```diff
@@ -888,6 +888,56 @@ function factory( dtype ) { // eslint-disable-line max-lines-per-function, stdli
+			v = buf[ GETTER ]( ( ( len - 1 ) * BYTES_PER_ELEMENT ) - ( i * BYTES_PER_ELEMENT ), this._isLE );
+			obuf[ SETTER ]( i * BYTES_PER_ELEMENT, v, this._isLE );
+		}
+
```
```suggestion
```

kgryte commented on lib/node_modules/@stdlib/array/fixed-endian-factory/test/test.to_reversed.js:
```diff
@@ -0,0 +1,120 @@
+	arr = new ctor( 'little-endian' );
+	out = arr.toReversed();
+
+	t.strictEqual(out.length, 0, 'returns expected value');
```
```suggestion
	t.strictEqual(out.length, 0, 'returns expected value' );
```

kgryte commented on lib/node_modules/@stdlib/array/fixed-endian-factory/test/test.to_reversed.js:
```diff
@@ -0,0 +1,120 @@
+
+	t.strictEqual( instanceOf( out, ctor ), true, 'returns expected value' );
+	t.notEqual( out, arr, 'returns a new instance' );
+	t.deepEqual( out, expected, 'returns expected value' );
```
You cannot use `deepEqual` as the returned instance uses accessors, not bracket notation, for accessing elements. You need to use `array/base/assert/has-same-values`.

jangnathan replied:
changed to:
```javascript
t.strictEqual( hasSameValues( out, expected ), true, 'returns expected value' );
```

kgryte reviewed (changes requested):
Thanks for working on this @nate10j. Left an initial round of comments.

jangnathan committed:
style: add missing space to array/fixed-endian-factory/benchmark

Co-authored-by: Athan <kgryte@gmail.com>
Signed-off-by: Nate <155411729+nate10j@users.noreply.github.com>

jangnathan committed:
fix: capitalise fail message in array/fixed-endian-factory/benchmark

Co-authored-by: Athan <kgryte@gmail.com>
Signed-off-by: Nate <155411729+nate10j@users.noreply.github.com>

jangnathan committed:
refactor: edit arr variable declaration in array/fixed-endian-factory/benchmark

Co-authored-by: Athan <kgryte@gmail.com>
Signed-off-by: Nate <155411729+nate10j@users.noreply.github.com>

jangnathan committed:
fix: edit fail message in array/fixed-endian-factory/benchmark

Co-authored-by: Athan <kgryte@gmail.com>
Signed-off-by: Nate <155411729+nate10j@users.noreply.github.com>

jangnathan committed:
docs: update JSDoc comment toReversed method in array/fixed-endian-factory

Co-authored-by: Athan <kgryte@gmail.com>
Signed-off-by: Nate <155411729+nate10j@users.noreply.github.com>

jangnathan committed:
docs: edit JSDoc comment to toReversed method in array/fixed-endian-factory

Co-authored-by: Athan <kgryte@gmail.com>
Signed-off-by: Nate <155411729+nate10j@users.noreply.github.com>

jangnathan committed:
docs: edit JSDoc comment to toReversed method in array/fixed-endian-factory

Co-authored-by: Athan <kgryte@gmail.com>
Signed-off-by: Nate <155411729+nate10j@users.noreply.github.com>

jangnathan committed:
perf: minimize arithmetic in array/fixed-endian-factory

Co-authored-by: Athan <kgryte@gmail.com>
Signed-off-by: Nate <155411729+nate10j@users.noreply.github.com>

jangnathan committed:
style: add missing space to array/fixed-endian-factory/test

Co-authored-by: Athan <kgryte@gmail.com>
Signed-off-by: Nate <155411729+nate10j@users.noreply.github.com>

style: add missing space to array/fixed-endian-factory/test

style: add missing space to test for toReversed method

jangnathan committed:
style: remove blank line for toReversed method in array/fixed-endian-factory

Co-authored-by: Athan <kgryte@gmail.com>
Signed-off-by: Nate <155411729+nate10j@users.noreply.github.com>

style: remove blank line for toReversed method in array/fixed-endian-factory

style: remove blank line for toReversed method in array/fixed-endian-factory

style: remove blank line in toReversed method

jangnathan committed:
style: remove blank line for toReversed method in array/fixed-endian-factory

Co-authored-by: Athan <kgryte@gmail.com>
Signed-off-by: Nate <155411729+nate10j@users.noreply.github.com>

style: remove blank line for toReversed method in array/fixed-endian-factory

style: remove blank line in toReversed method in array/fixed-endian-factory

style: remove blank line for toReversed method

jangnathan committed:
docs: remove duplicate JSDoc comment in array/fixed-endian-factory

jangnathan committed:
test: use hasSameValues method to test in array/fixed-endian-factory

jangnathan committed:
style: insert toReversed in alphabetic order in array/fixed-endian-factory

jangnathan commented:
@kgryte Thanks for the review, I have made required changes
